### PR TITLE
Update otel dependencies from v0.35.0 to v0.42.0

### LINKED ISSUE #109829: Add Tracing to the kubelet's read-only endpoint
### What would you like to be added?

Follow-up to https://github.com/kubernetes/kubernetes/pull/105126.  Discussed in https://github.com/kubernetes/kubernetes/pull/105126#discussion_r783142176.

As part of kubelet tracing beta, we should add tracing support to the kubelet's read-only endpoint, which currently does not have tracing.  It should use WithPublicEndpoint to prevent callers from influencing sampling decisions.

### Why is this needed?

Part of https://github.com/kubernetes/enhancements/issues/2831

### PR DESCRIPTION
**Pull request by frzifus**
#### What type of PR is this?

/kind feature
/sig node
/sig instrumentation

#### What this PR does / why we need it:
This pr updates:
- otel from v0.35.0 to v0.42.0
- contrib from v0.35.0 to v0.42.0
- grpc from v1.51.0 to v1.56.0

#### Which issue(s) this PR fixes:

Part of #109829

#### Special notes for your reviewer:

~Does it make sense to split the dependency update and the code change?~

I did split this up.

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

- KEP: https://github.com/kubernetes/enhancements/tree/master/keps/sig-instrumentation/647-apiserver-tracing

### CODE REVIEW

**Comment by liggitt** [comment]
> * [Untangle genproto / cloud.google.com/go dependencies #113366](https://github.com/kubernetes/kubernetes/issues/113366)
> 
> I think we're stuck with versions we have until we make a dent with this problem.
> 
> cc @liggitt

Yeah, @jpbetz, @cici37, and I are working with some of the folks maintaining genproto and `cloud.google.com` to get a plan in place for resolving https://github.com/kubernetes/kubernetes/issues/113366. Will update there soon, hopefully.

File: vendor/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc/config.go [hunk 1]
```diff
@@ -55,10 +57,22 @@ func newConfig(opts []Option) *config {
...
 		o.apply(c)
 	}
+
+	c.meter = c.MeterProvider.Meter(
+		instrumentationName,
+		metric.WithInstrumentationVersion(SemVersion()),
+		metric.WithSchemaURL(semconv.SchemaURL),
+	)
+	var err error
+	if c.rpcServerDuration, err = c.meter.Int64Histogram("rpc.server.duration", metric.WithUnit("ms")); err != nil {
```
Comment (liggitt) [inline vendor/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc/config.go:72]: this looks like it triggers the cardinality / memory issue reported in https://github.com/open-telemetry/opentelemetry-go-contrib/issues/3071

Comment (frzifus) [reply vendor/go.opentelemetry.io/contrib/instrumentation/google.golang.org/grpc/otelgrpc/config.go:72]: mh.. i will look into that.

**Comment by pacoxu** [comment]
The `pull-kubernetes-dependencies` may fail unexpectedly. I tried to fix it in #118707. Please rebase with the master after 118707 is merged.

**Comment by frzifus** [comment]
thx @pacoxu

**Comment by frzifus** [comment]
rebased @pacoxu

**Comment by dashpole** [comment]
Looks like this doesn't build:

`vendor/go.opentelemetry.io/otel/exporters/otlp/otlptrace/internal/otlpconfig/envconfig.go:84:13: undefined: envconfig.WithTLSConfig`

I haven't looked closely, but make sure you update all unstable otel dependencies to the same version.

**Comment by frzifus** [comment]
> Looks like this doesn't build: ...

Seems that fixed it. It was stuck at v1.10.0
```bash
./hack/pin-dependency.sh go.opentelemetry.io/otel/exporters/otlp/otlptrace/otlptracegrpc v1.16.0
```

But now the test seems to stuck and run out of time? :X 
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/117945/pull-kubernetes-node-e2e-containerd/1677218531566948352/#1:build-log.txt%3A3662

File: go.mod [hunk 2]
```diff
@@ -28,7 +28,7 @@ require (
 	github.com/cyphar/filepath-securejoin v0.2.3
 	github.com/docker/distribution v2.8.2+incompatible
 	github.com/docker/go-units v0.5.0
-	github.com/emicklei/go-restful/v3 v3.9.0
+	github.com/emicklei/go-restful/v3 v3.10.2
```
Comment (liggitt) [inline go.mod:31]: I'm pretty sure this bump breaks things, see https://github.com/kubernetes/kubernetes/pull/115067

Comment (frzifus) [reply go.mod:31]: I mean the [go-contrib instrumentation depends](https://github.com/open-telemetry/opentelemetry-go-contrib/blob/v1.16.0/instrumentation/github.com/emicklei/go-restful/otelrestful/go.mod#L8) on that. ^^

I hope I can have a deeper look on Friday. At least their changelogs do not look suspicious.
https://github.com/emicklei/go-restful/blob/v3/CHANGES.md

Comment (frzifus) [reply go.mod:31]: Whats the best way to debug this locally?

I did run the test locally and noticed that `TestCalculateLinuxResources` in `pkg/kubelet/kuberuntime` failed.
```diff
--- Expected
+++ Actual
@@ -9,3 +9,5 @@
  HugepageLimits: ([]*v1.HugepageLimit) <nil>,
- Unified: (map[string]string) <nil>,
+ Unified: (map[string]string) (len=1) {
+  (string) (len=16) "memory.oom.group": (string) (len=1) "1"
+ },
MemorySwapLimitInBytes: (int64) 0,
```

But in that case I assume its a local issue with my system settings :facepalm: 

https://github.com/kubernetes/kubernetes/blob/79c02ceb73f4d64f52a9d4c46785e4c7497493d9/pkg/kubelet/kuberuntime/kuberuntime_container_linux.go#L218-L226

**Comment by frzifus** [comment]
fixed by https://github.com/kubernetes/kubernetes/pull/121111

**Pull request closed without merging**